server/src/main/java/com/tinyfs/credentials/CredentialsObfuscator.java

```diff
@@ -74,16 +76,28 @@ public ClientCredentials deobfuscateClientCredentials(
   private <E> E deobfuscateProtobufCredentials(
       final byte[] obfuscatedCredentials,
       final ByteArrayParser<E> parseFunction) {
-    E credentials = null;
+    byte[] base64 = null;
+    try {
+      base64 = Base64.getDecoder().decode(obfuscatedCredentials);
+    } catch (IllegalArgumentException e) {
+      // Client sent non-base64 token string
+      throw new InvalidCredentialsException("Client gave non-Base64 token", e);
+    }
 
+    byte[] decryptedBase64 = null;
     try {
-      credentials =
-        parseFunction.parseFrom(decodeByteArray(fromBase64(obfuscatedCredentials)));
-    } catch (Exception e) {
-      throw new InvalidCredentialsException(e);
+      decryptedBase64 = decryptCipher.doFinal(base64);
+    } catch (IllegalBlockSizeException | BadPaddingException e) {
+      // Client sent unregistered token?
+      throw new InvalidCredentialsException("Client gave bad token", e);
     }
 
-    return credentials;
+    try {
+      return parseFunction.parseFrom(decryptedBase64);
+    } catch (Exception e) {
+      // ???
```

Comment:
this would probably be a mistake on our end if this is thrown: it means a client gave a message which was signed using our key but not a valid protobuf object to be parsed.